# Предобработка данных

- Автор: Кропотов Валентин
- Дата: 31.08.2025

### Цели и задачи проекта

- Провести предобработку данных.
- Получить нужный срез по годам: 2000-2013гг.
- Провести категоризацию по оценкам игроков и критиков.
- Выделить топ-7 платформ по количеству игр за данный период.

### Описание данных

Данные `new_games.csv` содержат информацию о продажах игр разных жанров и платформ, а также пользовательские и экспертные оценки игр:
- `Name` — название игры.
- `Platform` — название платформы.
- `Year of Release` — год выпуска игры.
- `Genre` — жанр игры.
- `NA sales` — продажи в Северной Америке (в миллионах проданных копий).
- `EU sales` — продажи в Европе (в миллионах проданных копий).
- `JP sales` — продажи в Японии (в миллионах проданных копий).
- `Other sales` — продажи в других странах (в миллионах проданных копий).
- `Critic Score` — оценка критиков (от 0 до 100).
- `User Score` — оценка пользователей (от 0 до 10).
- `Rating` — рейтинг организации ESRB (англ. Entertainment Software Rating Board). Эта ассоциация определяет рейтинг компьютерных игр и присваивает им подходящую возрастную категорию.

### Содержимое проекта

1. Загрузка данных и знакомство с ними
2. Проверка ошибок в данных и их предобработка
3. Фильтрация данных
4. Категоризация данных
5. Итоговый вывод

## Загрузка данных и знакомство с ними

- Загрузите необходимые библиотеки Python и данные датасета `/datasets/new_games.csv`.


In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('new_games.csv')

- Познакомьтесь с данными: выведите первые строки и результат метода `info()`.


In [3]:
# Знакомимся с данными:
df.info()
# Сохраним в переменную изначальное количество строк:
start_rows = df.shape[0]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Name             16954 non-null  object 
 1   Platform         16956 non-null  object 
 2   Year of Release  16681 non-null  float64
 3   Genre            16954 non-null  object 
 4   NA sales         16956 non-null  float64
 5   EU sales         16956 non-null  object 
 6   JP sales         16956 non-null  object 
 7   Other sales      16956 non-null  float64
 8   Critic Score     8242 non-null   float64
 9   User Score       10152 non-null  object 
 10  Rating           10085 non-null  object 
dtypes: float64(4), object(7)
memory usage: 1.4+ MB


In [4]:
# Первые 5 строк датасета:
df.head()

,Name,Platform,Year of Release,Genre,NA sales,EU sales,JP sales,Other sales,Critic Score,User Score,Rating
0,Wii Sports,Wii,2006.0,Sports,41.36,28.96,3.77,8.45,76.0,8,E
1,Super Mario Bros.,NES,1985.0,Platform,29.08,3.58,6.81,0.77,NaN,NaN,NaN
2,Mario Kart Wii,Wii,2008.0,Racing,15.68,12.76,3.79,3.29,82.0,8.3,E
3,Wii Sports Resort,Wii,2009.0,Sports,15.61,10.93,3.28,2.95,80.0,8,E
4,Pokemon Red/Pokemon Blue,GB,1996.0,Role-Playing,11.27,8.89,10.22,1.00,NaN,NaN,NaN


In [5]:
#  Cлучайные 5 строк датасета:
df.sample(5, random_state=1)

,Name,Platform,Year of Release,Genre,NA sales,EU sales,JP sales,Other sales,Critic Score,User Score,Rating
9596,Diablo II,PC,2000.0,Role-Playing,0.01,0.1,0.0,0.02,88.0,8.8,M
1253,Call of Duty: Modern Warfare: Reflex Edition,Wii,2009.0,Shooter,0.94,0.43,0.0,0.14,NaN,NaN,NaN
6405,Rory McIlroy PGA Tour,XOne,2015.0,Action,0.16,0.09,0.0,0.02,60.0,4.3,E
10264,WWE All Stars,PS2,2011.0,Fighting,0.05,0.04,0.0,0.01,NaN,3.4,T
8157,Zumba Fitness Rush,X360,2012.0,Sports,0.00,0.16,0.0,0.02,73.0,6.2,E10+


In [6]:
# Узнаеем количество пропусков:
print(f'Количество пропусков в абсолютных значениях:\n{df.isna().sum()}')
# И долю пропусков от общего числа строк:
print(f'\nДоля пропусков от общего числа строк:\n{df.isna().sum()/len(df)}')

Количество пропусков в абсолютных значениях:
Name                  2
Platform              0
Year of Release     275
Genre                 2
NA sales              0
EU sales              0
JP sales              0
Other sales           0
Critic Score       8714
User Score         6804
Rating             6871
dtype: int64

Доля пропусков от общего числа строк:
Name               0.000118
Platform           0.000000
Year of Release    0.016218
Genre              0.000118
NA sales           0.000000
EU sales           0.000000
JP sales           0.000000
Other sales        0.000000
Critic Score       0.513918
User Score         0.401274
Rating             0.405225
dtype: float64


- Сделайте вывод о полученных данных: данные какого объёма вам предоставили, соответствуют ли они описанию, встречаются ли в них пропуски, используются ли верные типы данных.
- Отметьте другие особенности данных, которые вы обнаружили и на которые стоит обратить внимание при предобработке. Например, вы можете проверить названия столбцов: все ли названия отражают содержимое данных и прописаны в удобном для работы виде.

**Объём данных и пропуски:**
- В датасете 11 колонок и 16 956 строк.
- Названия столбцов и данные в них соответствуют описанию, но названия следует привести к стилю snake case. И я бы еще поменял название столбца `Rating` на `rating_esrb`, чтобы стало более очевидно, что за информация в нем хранится.
- Два пропуска присутствуют в колонках `Name` и `Genre`. Возможно это даже одни и те же строки, стоит проверить.
- В столбце `Year of Release` 275 пропусков, что составляет примерно 1.6% от общего числа строк.
- Основные пропуски находятся в столбцах `Critic Score` - оценка критиков (доля пропусков-51%), `User Score` - оценка пользователей (40% пропусков) и `Rating` - рейтинг организации ESRB (40% пропусков).

**Типы данных:**
- У явно числовых столбцов `EU sales`, `JP sales` и `User Score` тип данных `object`. Нужно выяснять почему.
- `Year of Release` следует привести к целочисленному типу. Мне кажется нет смысла преобразовывать этот столбец в дату и время.
- Как я понимаю рейтинг критиков в столбце `Critic Score` тоже не имеет дробной части, можно привести его к типу `integer`.
- У остальных столбцов тип данных адекватен находящимся в них данных.

---

## Проверка ошибок в данных и их предобработка


### Названия, или метки, столбцов датафрейма

- Выведите на экран названия всех столбцов датафрейма и проверьте их стиль написания.
- Приведите все столбцы к стилю snake case. Названия должны быть в нижнем регистре, а вместо пробелов — подчёркивания.

In [7]:
# Проверяем стиль написания названий столбцов
df.columns

Index(['Name', 'Platform', 'Year of Release', 'Genre', 'NA sales', 'EU sales',
       'JP sales', 'Other sales', 'Critic Score', 'User Score', 'Rating'],
      dtype='object')

- Названия столбцов представлены в обычном стиле. Для удобства стоит их привести к стилю snake case.

In [8]:
# Переводим в нижний регистр и заменяем пробелы нижним подчеркиванием:
df.columns = df.columns.str.lower().str.replace(' ', '_')
df.columns

Index(['name', 'platform', 'year_of_release', 'genre', 'na_sales', 'eu_sales',
       'jp_sales', 'other_sales', 'critic_score', 'user_score', 'rating'],
      dtype='object')

- Переименовываем последний столбец для большей ясности, что в нем хранится.

In [9]:
# Используем метод .rename для переименовывания:
df.rename(columns={'rating':'rating_esrb'}, inplace=True)
df.columns

Index(['name', 'platform', 'year_of_release', 'genre', 'na_sales', 'eu_sales',
       'jp_sales', 'other_sales', 'critic_score', 'user_score', 'rating_esrb'],
      dtype='object')

### Типы данных

- Если встречаются некорректные типы данных, предположите их причины.
- При необходимости проведите преобразование типов данных. Помните, что столбцы с числовыми данными и пропусками нельзя преобразовать к типу `int64`. Сначала вам понадобится обработать пропуски, а затем преобразовать типы данных.

- `eu_sales`, `jp_ sales` и `user_score` имеют тип `object` скорее всего из-за строковых значений внутри столбцов.
- `year_of_release` и `critic_score` следует привести к типу `integer`.

In [10]:
df['user_score'].unique()

array(['8', nan, '8.3', '8.5', '6.6', '8.4', '8.6', '7.7', '6.3', '7.4',
       '8.2', '9', '7.9', '8.1', '8.7', '7.1', '3.4', '5.3', '4.8', '3.2',
       '8.9', '6.4', '7.8', '7.5', '2.6', '7.2', '9.2', '7', '7.3', '4.3',
       '7.6', '5.7', '5', '9.1', '6.5', 'tbd', '8.8', '6.9', '9.4', '6.8',
       '6.1', '6.7', '5.4', '4', '4.9', '4.5', '9.3', '6.2', '4.2', '6',
       '3.7', '4.1', '5.8', '5.6', '5.5', '4.4', '4.6', '5.9', '3.9',
       '3.1', '2.9', '5.2', '3.3', '4.7', '5.1', '3.5', '2.5', '1.9', '3',
       '2.7', '2.2', '2', '9.5', '2.1', '3.6', '2.8', '1.8', '3.8', '0',
       '1.6', '9.6', '2.4', '1.7', '1.1', '0.3', '1.5', '0.7', '1.2',
       '2.3', '0.5', '1.3', '0.2', '0.6', '1.4', '0.9', '1', '9.7'],
      dtype=object)

In [11]:
# pd.to_numeric(df['user_score'])
# ValueError: Unable to parse string "tbd" at position 119

- Аббревиатура *tbd* означает *"To Be Decided" (будет решено)*. Из-за этой строки весь столбец стал типом `object`. В нашем случае такое значение означает пропуск, поскольку рейтинга в данных нет. Значит можно заменить эти значения с помощью параметра метода to_numeric - errors = 'coerce'

In [12]:
# Заменяем нечисловые значения пропусками и делаем тип данных более компактным:
df['user_score'] = pd.to_numeric(df['user_score'], errors = 'coerce', downcast = 'float')
df['user_score'].dtypes

dtype('float32')

- В числовых столбцах могут встретиться строковые значения, например `unknown` или другие. Приводите такие столбцы к числовому типу данных, заменив строковые значения на пропуски.

In [13]:
# df['eu_sales'].astype('float64')
# df['jp_sales'].astype('float64')
# ValueError: could not convert string to float: 'unknown'

- Для нас значение 'unknown' равносильно пропуску 'NaN'. Алгоритм тот же, что и в случае столбца `user_score`.

In [14]:
df['jp_sales'] = pd.to_numeric(df['jp_sales'], errors = 'coerce', downcast = 'float')
df['jp_sales'].dtypes

dtype('float32')

In [15]:
df['eu_sales'] = pd.to_numeric(df['eu_sales'], errors = 'coerce', downcast = 'float')
df['eu_sales'].dtypes

dtype('float32')

- Приведем столбцы `year_of_release` и `critic_score` к целочисленному типу, но сначала проверим уникальные значения в них.

In [16]:
df['critic_score'].unique()

array([76., nan, 82., 80., 89., 58., 87., 91., 61., 97., 95., 77., 88.,
       83., 94., 93., 85., 86., 98., 96., 90., 84., 73., 74., 78., 92.,
       71., 72., 68., 62., 49., 67., 81., 66., 56., 79., 70., 59., 64.,
       75., 60., 63., 69., 50., 25., 42., 44., 55., 48., 57., 29., 47.,
       65., 54., 20., 53., 37., 38., 33., 52., 30., 32., 43., 45., 51.,
       40., 46., 39., 34., 35., 41., 36., 28., 31., 27., 26., 19., 23.,
       24., 21., 17., 22., 13.])

In [17]:
df['year_of_release'].unique()

array([2006., 1985., 2008., 2009., 1996., 1989., 1984., 2005., 1999.,
       2007., 2010., 2013., 2004., 1990., 1988., 2002., 2001., 2011.,
       1998., 2015., 2012., 2014., 1992., 1997., 1993., 1994., 1982.,
       2016., 2003., 1986., 2000.,   nan, 1995., 1991., 1981., 1987.,
       1980., 1983.])

- Все значения являются числами с нулевой дробной частью. Можно смело менять тип данных.

In [18]:
# Оценку критиков меняем на Int8, потому что значение не может быть больше ста и в столбце присутствуют пропуски
# Год релиза меняем на Int16, потому что значения больше 127 (Int8 не подойдет) и присутствуют пропуски
df = df.astype({'critic_score':'Int8', 'year_of_release':'Int16'})
df[['critic_score', 'year_of_release']].dtypes

critic_score        Int8
year_of_release    Int16
dtype: object

- Так же можно "облегчить" `na_sales` и `other_sales`.

In [19]:
df[['na_sales', 'other_sales']] = df[['na_sales', 'other_sales']].astype('float32')
df[['na_sales', 'other_sales']].dtypes

na_sales       float32
other_sales    float32
dtype: object

### Наличие пропусков в данных

- Посчитайте количество пропусков в каждом столбце в абсолютных и относительных значениях.


In [20]:
# Узнаеем количество пропусков:
print(f'Количество пропусков в абсолютных значениях:\n{df.isna().sum()}')
# И долю пропусков от общего числа строк:
print(f'\nДоля пропусков от общего числа строк в процентах:\n{df.isna().mean()*100}')

Количество пропусков в абсолютных значениях:
name                  2
platform              0
year_of_release     275
genre                 2
na_sales              0
eu_sales              6
jp_sales              4
other_sales           0
critic_score       8714
user_score         9268
rating_esrb        6871
dtype: int64

Доля пропусков от общего числа строк в процентах:
name                0.011795
platform            0.000000
year_of_release     1.621845
genre               0.011795
na_sales            0.000000
eu_sales            0.035386
jp_sales            0.023590
other_sales         0.000000
critic_score       51.391838
user_score         54.659118
rating_esrb        40.522529
dtype: float64


- Изучите данные с пропущенными значениями. Напишите промежуточный вывод: для каких столбцов характерны пропуски и сколько их. Предположите, почему пропуски могли возникнуть. Укажите, какие действия с этими данными можно сделать и почему.


- Два пропуска присутствуют в колонках `name` и `genre`. Возможно это даже одни и те же строки, стоит проверить.
- В столбце `year_of_release` 275 пропусков, что составляет примерно 1.6% от общего числа строк. Пропуски выглядят как абсолютно случайными (MCAR). Возможно эти данные можно восстановить из открытых источников, но у нас нет столько времени. Думаю, можно их игнорировать.
- Основные пропуски находятся в столбцах `critic_score` - оценка критиков (доля пропусков - 51,4%), `user_score` - оценка пользователей (54,7% пропусков) и `rating_esrb` - рейтинг организации ESRB (40,5% пропусков). При этом оценка критиков и оценка игроков может отсутствовать, помимо технических причин, в виду банальной малоизвестности игры. Так же на наличие пропусков влияет источник данных для датасета, множество игроков оставляют свои оценки на разных сайтах-агрегаторах, либо вообще не делятся своим мнением. Судя по данным информация об оценках взята с сайта Metacritic и далеко не все геймеры оставляют там свои оценки, много внимания получают только популярные игры. Или же игра может быть старой, что, соответственно, снижает шанс, что люди, игравшие в нее, пойдут спустя много лет ставить ей оценку. В общем причин может быть много. Причиной пропусков в столбце `rating_esrb`, помимо вышеупомянутых, является то, что игра могла просто не попасть "на стол" к ESRB. Эта организация занимается определением рейтингов в США и Канаде, а игра в датасете могла и не выходить официальо на эти рынки.


#### **Пропуски в `name` и `genre`:**

In [21]:
# Проверяем пропуски в столбцах name и genre:
df[df['name'].isna()]

,name,platform,year_of_release,genre,na_sales,eu_sales,jp_sales,other_sales,critic_score,user_score,rating_esrb
661,NaN,GEN,1993,NaN,1.78,0.53,0.00,0.08,<NA>,NaN,NaN
14439,NaN,GEN,1993,NaN,0.00,0.00,0.03,0.00,<NA>,NaN,NaN


- Да, это действительно оказались одни и те же строки, но нас интересуют игры, которые вышли в релиз в период с 2000 по 2013. Эти пропуски можно игнорировать.

#### **Пропуски в `year_of_release`**

In [22]:
# Смотрим случайные 10 строк с пропусками в столбце year_of_release:
df[df['year_of_release'].isna()].sample(10, random_state=4)

,name,platform,year_of_release,genre,na_sales,eu_sales,jp_sales,other_sales,critic_score,user_score,rating_esrb
10831,Happy Feet Two,DS,<NA>,Action,0.08,0.01,0.00,0.01,<NA>,NaN,E
14319,Housekeeping,DS,<NA>,Action,0.00,0.00,0.04,0.00,<NA>,NaN,NaN
9508,Transworld Surf,XB,<NA>,Sports,0.10,0.03,0.00,0.00,76,NaN,T
13747,RollerCoaster Tycoon,PC,<NA>,Strategy,0.02,0.02,0.00,0.01,<NA>,NaN,E
5727,All-Star Baseball 2005,PS2,<NA>,Sports,0.16,0.12,0.00,0.04,72,8.6,E
3388,The Golden Compass,Wii,<NA>,Action,0.26,0.28,0.00,0.07,35,6.8,E10+
1714,Hitman 2: Silent Assassin,XB,<NA>,Action,0.76,0.38,0.00,0.05,84,8.0,M
7452,Ghostbusters II,2600,<NA>,Action,0.20,0.01,0.00,0.00,<NA>,NaN,NaN
16584,Agarest Senki: Re-appearance,PS3,<NA>,Role-Playing,0.00,0.00,0.01,0.00,<NA>,NaN,NaN
3450,Madden NFL 06,X360,<NA>,Sports,0.54,0.00,0.01,0.03,74,4.9,E


- В этой рандомной выборке у всех игр можно узнать год релиза. А в спортивных играх, например, обычно год релиза = год в названии(если он есть) - 1.

- Пропуски выглядят абсолютно случайными (MCAR). Даже у очень популярной игры Hitman 2: Silent Assassin с суммарными продажами более миллиона копий в датасете отсутствует год релиза, т.е. дело точно не в безызвестности игр. Думаю, что в пропусках выноваты технические неполадки или другие случайности.

- Можно еще посмотреть как ведут себя пропуски при группировке по платформе и жанру, но сначала избавимся от неявных дубликатов, они будут мешать группировке.

In [23]:
df['genre'].unique()

array(['Sports', 'Platform', 'Racing', 'Role-Playing', 'Puzzle', 'Misc',
       'Shooter', 'Simulation', 'Action', 'Fighting', 'Adventure',
       'Strategy', nan, 'MISC', 'ROLE-PLAYING', 'RACING', 'ACTION',
       'SHOOTER', 'FIGHTING', 'SPORTS', 'PLATFORM', 'ADVENTURE',
       'SIMULATION', 'PUZZLE', 'STRATEGY'], dtype=object)

In [24]:
df['platform'].unique()

array(['Wii', 'NES', 'GB', 'DS', 'X360', 'PS3', 'PS2', 'SNES', 'GBA',
       'PS4', '3DS', 'N64', 'PS', 'XB', 'PC', '2600', 'PSP', 'XOne',
       'WiiU', 'GC', 'GEN', 'DC', 'PSV', 'SAT', 'SCD', 'WS', 'NG', 'TG16',
       '3DO', 'GG', 'PCFX'], dtype=object)

- В названии жанров есть неявные дубликаты. Нужно привести их к единому стилю написания.

In [25]:
# Приведем значения в стобцах genre к нижнему регистру, а platform к верхнему
df['genre'] = df['genre'].str.lower()

In [26]:
# Создадим функции для вычисления суммарного кол-ва пропусков и доли пропусков в процентах:
def sum_nans(x):
    return x.isna().sum()
def mean_nans(x):
    return x.isna().mean()*100

In [27]:
df.groupby('genre')['year_of_release'].agg([sum_nans, mean_nans]).sort_values(by='mean_nans',ascending=False)

,sum_nans,mean_nans
genre,,
shooter,27,2.005944
simulation,17,1.923077
action,65,1.901697
puzzle,11,1.864407
sports,43,1.810526
racing,23,1.806756
misc,29,1.636569
strategy,11,1.591896
fighting,12,1.392111


In [28]:
df.groupby('platform')['year_of_release'].agg([sum_nans, mean_nans]).sort_values(by='mean_nans',ascending=False)

,sum_nans,mean_nans
platform,,
2600,17,12.592593
Wii,35,2.611940
XB,21,2.502980
X360,32,2.498048
GC,14,2.486679
PS3,25,1.845018
PC,18,1.818182
PS2,35,1.598904
3DS,8,1.509434


- Пропуски по жанрам распределяются довольно равномерно, а вот по платформам не совсем. Больше всего доля пропусков на платформе 2600. Это приставка Atari 2600, игры для нее выходили до нужного нам диапазона лет.
- Из платформ в топе по доле пропусков есть как популярные, так и не очень. За пределами Atari 2600 я не знаю, с чем может быть связана разница в количестве и доле пропусков.
  
  **Вывод:**
- Таким образом получается, что пропуски в столбце `year_of_release` как-то связаны с платформой, на которой выходила игра. При этом срок жизни и выпуск выдеоигр на некоторые платформы почти точно совпадает с искомым диапазоном, но этой информации в датасете нет, значит это пропуски MAR. Можно заполнить строки с пропусками в платформе PS2, PSP, DS, XB, GC - игры на них выходили как раз примерно с 2000 по 2013. Остальные строки с пропусками в столбце `year_of_release` можно либо удалить, либо заполнить модой по категории платформы. Тут я бы уточнил, что нам важнее: точно не допустить лишних игр в финальную выборку (тогда лучше удалить), либо допустить небольшое количество игр, которые на самом деле не выходили с 2000 по 2013.

#### **Пропуски в `critic_score`, `user_score` и `rating_esrb`**

- Рассмотрим случайную выборку из строк с пропусками в этих столбцах.

In [29]:
df[df['critic_score'].isna()].sample(10, random_state=1)

,name,platform,year_of_release,genre,na_sales,eu_sales,jp_sales,other_sales,critic_score,user_score,rating_esrb
4891,Mystery Dungeon: Shiren the Wanderer,SNES,1995,role-playing,0.00,0.00,0.40,0.00,<NA>,NaN,NaN
14786,Galaxy Angel II: Mugen Kairou no Kagi,PS2,2007,strategy,0.00,0.00,0.03,0.00,<NA>,NaN,NaN
11899,Codename: Kids Next Door: Operation S.O.D.A.,GBA,2004,platform,0.05,0.02,0.00,0.00,<NA>,NaN,E
13158,Minna no Ennichi,3DS,2012,action,0.00,0.00,0.05,0.00,<NA>,NaN,NaN
7865,Just Dance: Greatest Hits,Wii,2012,misc,0.14,0.04,0.00,0.01,<NA>,NaN,E10+
14625,Totally Spies!,GBA,2005,platform,0.02,0.01,0.00,0.00,<NA>,NaN,NaN
11640,The Sims 3: Fast Lane Stuff,PC,2010,simulation,0.01,0.05,0.00,0.02,<NA>,5.4,T
16473,Fernbus Simulator,PC,2016,simulation,0.00,0.01,0.00,0.00,<NA>,NaN,NaN
11493,Armorines: Project S.W.A.R.M.,PS,2000,shooter,0.05,0.03,0.00,0.01,<NA>,NaN,NaN
14348,PriPara Mezameyo! Megami no Dress Design,3DS,2016,action,0.00,0.00,0.03,0.00,<NA>,NaN,NaN


In [30]:
df[df['user_score'].isna()].sample(10, random_state=1)

,name,platform,year_of_release,genre,na_sales,eu_sales,jp_sales,other_sales,critic_score,user_score,rating_esrb
16411,Fab 5 Soccer,DS,2008,sports,0.01,0.00,0.00,0.00,<NA>,NaN,E
16277,Tiger & Bunny: Hero's Day,PSP,2013,action,0.00,0.00,0.02,0.00,<NA>,NaN,NaN
16191,V-Rally 3,XB,2003,racing,0.01,0.00,0.00,0.00,76,NaN,E
13218,Piglet's Big Game,GBA,2003,platform,0.04,0.01,0.00,0.00,<NA>,NaN,NaN
13370,Codename: Kids Next Door: Operation V.I.D.E.O....,GC,2005,platform,0.04,0.01,0.00,0.00,58,NaN,E10+
13681,Looney Tunes: Cartoon Conductor,DS,2008,misc,0.04,0.00,0.00,0.00,67,NaN,E
6411,Jikkyou Powerful Pro Yakyuu 5,N64,1998,sports,0.00,0.00,0.27,0.00,<NA>,NaN,NaN
16065,Clock Zero: Shuuen no Ichibyou - ExTime,PSV,2015,action,0.00,0.00,0.02,0.00,<NA>,NaN,NaN
4624,Big Strike Bowling,PS,2003,sports,0.23,0.16,0.00,0.03,<NA>,NaN,NaN
10228,Queen's Blade: Spiral Chaos,PSP,2009,role-playing,0.00,0.00,0.11,0.00,<NA>,NaN,NaN


In [31]:
df[df['rating_esrb'].isna()].sample(10, random_state=7)

,name,platform,year_of_release,genre,na_sales,eu_sales,jp_sales,other_sales,critic_score,user_score,rating_esrb
15794,Amatsumi Sora ni! Kumo no Hatate ni,PSV,2015,adventure,0.00,0.00,0.02,0.00,<NA>,NaN,NaN
9707,Sengoku Basara 3 Utage,Wii,2011,action,0.00,0.00,0.13,0.00,<NA>,NaN,NaN
12694,Summon Night 2,DS,2008,role-playing,0.00,0.00,0.06,0.00,<NA>,NaN,NaN
1086,Fire Emblem Fates,3DS,2015,role-playing,0.81,0.23,0.52,0.11,<NA>,NaN,NaN
634,Final Fantasy V,SNES,1992,role-playing,0.00,0.00,2.43,0.02,<NA>,NaN,NaN
343,Star Wars Battlefront (2015),XOne,2015,shooter,2.04,1.28,0.02,0.32,<NA>,NaN,NaN
14893,Battlefield 1942: The WW II Anthology,PC,2004,shooter,0.01,0.01,0.00,0.00,<NA>,NaN,NaN
14712,Gotouchi Tetsudou: Gotouchi Chara to Nihon Zen...,WiiU,2014,misc,0.00,0.00,0.03,0.00,<NA>,NaN,NaN
5959,Macross Digital Mission VF-X,PS,1997,simulation,0.00,0.00,0.28,0.02,<NA>,NaN,NaN
11499,Thomas & Friends: Hero of the Rails,DS,2010,misc,0.00,0.07,0.00,0.01,<NA>,NaN,NaN


- Во всех трех случаях глаз цепляется за низкие показатели продаж, т.е. низкой популярности игры. Кроме Star Wars Battlefront (2015), который вышел в 2015 и с суммарными продажами больше 3 миллионов копий, но при этом у него нет ни оценки критиков, ни оценки игроков, ни рейтинга ESRB. Скорее всего это значит, что в пропусках по этим столбцам есть доля совершенно случайных (MCAR).

**`rating_esrb`**
- `rating_esrb` нужно проверить группировкой по платформам. В датасете этой информации нет, но я знаю, что многие платформы были либо непопулярны, либо вообще не выходили на рынки США и Канады, где и работает организация ESRB.

In [32]:
df.groupby('platform')['rating_esrb'].agg([sum_nans, mean_nans]).sort_values(by='mean_nans',ascending=False)

,sum_nans,mean_nans
platform,,
2600,135,100.000000
3DO,3,100.000000
GB,98,100.000000
NES,100,100.000000
N64,323,100.000000
GG,1,100.000000
GEN,29,100.000000
PCFX,1,100.000000
NG,12,100.000000


**Вывод:**
- Гипотеза подтвердилась - больше всего доля пропусков на платформах, которые либо существовали до основания организации ESRB, либо не были популярны/не выходили на рынок в США и Канаде. Меньше всего пропусков на популярных в США платформах. Это значит, что пропуски означают отсутствие значения. Но при этом некоторые пропуски являются абсолютно случайными, т.е. в природе они существуют, но у нас в датасете их нет.
- Учитывая количество и тип пропусков лучше всего заполнить их индикатором, который будет сообщать об отсутствии рейтинга в датасете.

**`user_score` и `critic_score`**
- Сгруппируем пропуски в оценках критиков и игроков по жанру игр.

In [33]:
df.groupby('genre')['user_score'].agg([sum_nans, mean_nans]).sort_values(by='mean_nans',ascending=False)

,sum_nans,mean_nans
genre,,
puzzle,460,77.966102
adventure,1020,77.097506
misc,1326,74.830700
simulation,535,60.520362
strategy,376,54.413893
fighting,457,53.016241
sports,1259,53.010526
platform,472,52.212389
role-playing,767,50.593668


In [34]:
df.groupby('genre')['critic_score'].agg([sum_nans, mean_nans]).sort_values(by='mean_nans',ascending=False)

,sum_nans,mean_nans
genre,,
adventure,996,75.283447
misc,1240,69.977427
puzzle,364,61.694915
simulation,527,59.615385
strategy,385,55.716353
fighting,448,51.972158
role-playing,773,50.989446
sports,1165,49.052632
platform,401,44.358407


- Так же можно посмотреть зависимость от года выхода игры

In [35]:
df.groupby('year_of_release')['user_score'].agg([sum_nans, mean_nans]).sort_values(by='mean_nans',ascending=False)

,sum_nans,mean_nans
year_of_release,,
1980,9,100.000000
1981,46,100.000000
1982,37,100.000000
1983,18,100.000000
1984,14,100.000000
1986,22,100.000000
1987,17,100.000000
1990,16,100.000000
1989,17,100.000000


In [36]:
df.groupby('year_of_release')['critic_score'].agg([sum_nans, mean_nans]).sort_values(by='mean_nans',ascending=False)

,sum_nans,mean_nans
year_of_release,,
1980,9,100.000000
1981,46,100.000000
1982,37,100.000000
1983,18,100.000000
1984,14,100.000000
1986,22,100.000000
1987,17,100.000000
1990,16,100.000000
1989,17,100.000000


**Вывод:**
- Как можно заметить есть зависимость количества пропусков от жанровой принадлежности игр и года релиза. При этом в абсолютных значениях количество пропусков в оценках игроков и критиков очень схожи. Мне кажется тут дело в непопулярности игр и жанров, а так же "возраст" игры, т.е. отсутвие оценок. Мы видим явный провал до 2000-х и большое количество пропусков в 2010-х. Возможно это связано с тем, что примерно в 2010-х индустрия выросла и появилось огромное количество игр, на которые количества игроков и критиков просто не хватило. Но не стоит забывать про совершенно случайные пропуски, которые точно есть, как мы успели убедиться ранее.
- Учитывая количество и характер пропусков предлагаю заменить их на индикатор.

- Обработайте пропущенные значения. Для каждого случая вы можете выбрать оптимальный, на ваш взгляд, вариант: заменить на определённое значение, оставить как есть или удалить.
- Если вы решите заменить пропуски на значение-индикатор, то убедитесь, что предложенное значение не может быть использовано в данных.
- Если вы нашли пропуски в данных с количеством проданных копий игры в том или ином регионе, их можно заменить на среднее значение в зависимости от названия платформы и года выхода игры.

#### **Обработка пропусков**

- **`year_of_release`**

In [37]:
# Заполним пропустки в year_of_release по платформам PS2, PSP, DS, XB, GC их модой:
required_platforms = ['PS2', 'PSP', 'DS', 'XB', 'GC']

# Создадим маску для строк с нужными платформами:
mask = df['platform'].isin(required_platforms)

# Заполнение пропусков только для нужных платформ:
df.loc[mask, 'year_of_release'] = df[mask].groupby('platform')['year_of_release'].transform(
    lambda x: x.fillna(x.mode()[0] if not x.mode().empty else np.nan)
)
print(f'Оставшееся количество строк с пропусками по столбцу year_of_release: {df['year_of_release'].isna().sum()}')
print(f'\nПроцент оставшихся пропусков: {df['year_of_release'].isna().mean()*100}')

Оставшееся количество строк с пропусками по столбцу year_of_release: 158

Процент оставшихся пропусков: 0.9318235432885115


- Остальные 158 пропусков можно удалить.

In [38]:
df = df.dropna(subset='year_of_release')

- **`critic_score` и `user_score`**

  Заменяем пропуски на индикатор "-1", потому что оценка не может быть отрицательной.

In [39]:
df[['critic_score', 'user_score']] = df[['critic_score', 'user_score']].fillna(-1)

- **`rating_esrb`**

 Заменяем пропуски на "no rating", потому у нас нет информации о рейтинге.

In [40]:
df['rating_esrb'] = df['rating_esrb'].fillna('no rating')

- **`eu_sales` и `jp_sales`**

Заменяем пропуски на среднее значение в зависимости от платформы и года выхода игры.

In [41]:
df['eu_sales'] = df['eu_sales'].fillna(df.groupby(['platform', 'year_of_release'])['eu_sales'].transform('mean'))

In [42]:
df['jp_sales'] = df['jp_sales'].fillna(df.groupby(['platform', 'year_of_release'])['jp_sales'].transform('mean'))

In [43]:
# Проверяем, что получилось:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 16798 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16796 non-null  object 
 1   platform         16798 non-null  object 
 2   year_of_release  16798 non-null  Int16  
 3   genre            16796 non-null  object 
 4   na_sales         16798 non-null  float32
 5   eu_sales         16798 non-null  float32
 6   jp_sales         16798 non-null  float32
 7   other_sales      16798 non-null  float32
 8   critic_score     16798 non-null  Int8   
 9   user_score       16798 non-null  float32
 10  rating_esrb      16798 non-null  object 
dtypes: Int16(1), Int8(1), float32(5), object(4)
memory usage: 1.0+ MB


- Осталось по 2 пропуска в `name` и `genre`, которые я решил игнорировать, потому что в финальный датафрейм они не войдут из-за года выхода.

### Явные и неявные дубликаты в данных

- Изучите уникальные значения в категориальных данных, например с названиями жанра игры, платформы, рейтинга и года выпуска. Проверьте, встречаются ли среди данных неявные дубликаты, связанные с опечатками или разным способом написания.
- При необходимости проведите нормализацию данных с текстовыми значениями. Названия или жанры игр можно привести к нижнему регистру, а названия рейтинга — к верхнему.

- **Столбец `genre` мы уже привели к одинаковому написанию**
- **В столбце `platform` не было дубликатов**

- **Проверяем столбец `rating_esrb`**

In [44]:
# Проверяем уникальные значения столбца rating_esrb:
df['rating_esrb'].unique()

array(['E', 'no rating', 'M', 'T', 'E10+', 'K-A', 'AO', 'EC', 'RP'],
      dtype=object)

- Рейтинг "K-A" и "E" означают одно и то же. "K-A" - это более раннее название, сейчас оно не используется.
- Можно заменить "K-A" на "E".

In [45]:
df['rating_esrb'] = df['rating_esrb'].str.replace('K-A', 'E')

In [46]:
df['rating_esrb'].unique()

array(['E', 'no rating', 'M', 'T', 'E10+', 'AO', 'EC', 'RP'], dtype=object)

- **Столбец `year_of_release` мы тоже уже проверяли**

- **Можно еще привести `name` к нижнему регистру на всякий случай**

In [47]:
df['name'] = df['name'].str.lower()

- После того как нормализуете данные и устраните неявные дубликаты, проверьте наличие явных дубликатов в данных.

In [48]:
# Выводим явные дубликаты на основе всего датасета:
df[df.duplicated()]

,name,platform,year_of_release,genre,na_sales,eu_sales,jp_sales,other_sales,critic_score,user_score,rating_esrb
268,batman: arkham asylum,PS3,2009,action,2.24,1.31,0.07,0.61,91,8.9,T
368,james bond 007: agent under fire,PS2,2001,shooter,1.90,1.13,0.10,0.41,72,7.9,T
717,god of war: ascension,PS3,2013,action,1.23,0.63,0.04,0.35,80,7.5,M
823,wipeout: the game,Wii,2009,misc,1.94,0.00,0.00,0.12,-1,-1.0,no rating
848,rayman raving rabbids: tv party,Wii,2008,misc,0.72,1.08,0.00,0.23,73,7.7,E10+
...,...,...,...,...,...,...,...,...,...,...,...
16671,fullmetal alchemist: prince of the dawn,Wii,2009,adventure,0.00,0.00,0.01,0.00,-1,-1.0,no rating
16753,routes pe,PS2,2007,adventure,0.00,0.00,0.01,0.00,-1,-1.0,no rating
16799,transformers: prime,Wii,2012,action,0.00,0.01,0.00,0.00,-1,-1.0,no rating
16912,metal gear solid v: the definitive experience,XOne,2016,action,0.01,0.00,0.00,0.00,-1,-1.0,M


- Напишите промежуточный вывод: укажите количество найденных дубликатов и действия по их обработке.

In [49]:
print(f'Всего найдено {df[df.duplicated()].shape[0]} строк дубликатов.')

Всего найдено 237 строк дубликатов.


- Найденные полные дубликаты можно удалить.

In [50]:
df.drop_duplicates(inplace=True)
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 16561 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16559 non-null  object 
 1   platform         16561 non-null  object 
 2   year_of_release  16561 non-null  Int16  
 3   genre            16559 non-null  object 
 4   na_sales         16561 non-null  float32
 5   eu_sales         16561 non-null  float32
 6   jp_sales         16561 non-null  float32
 7   other_sales      16561 non-null  float32
 8   critic_score     16561 non-null  Int8   
 9   user_score       16561 non-null  float32
 10  rating_esrb      16561 non-null  object 
dtypes: Int16(1), Int8(1), float32(5), object(4)
memory usage: 1.0+ MB


- В процессе подготовки данных вы могли что-либо удалять, например строки с пропусками или ошибками, дубликаты и прочее. В этом случае посчитайте количество удалённых строк в абсолютном и относительном значениях.

In [51]:
print(f'''Общее количество удаленных строк: {start_rows - df.shape[0]}.
Доля удаленных строк относительно изначального количества: {(start_rows - df.shape[0])/start_rows}.''')

Общее количество удаленных строк: 395.
Доля удаленных строк относительно изначального количества: 0.023295588582212787.


- После проведения предобработки данных напишите общий промежуточный вывод.

- В датасете было обнаружено много пропусков в столбцах `critic_score`, `user_score` и `rating_esrb`, которые было решено заменить на индикаторы.
- Некоторые пропуски в столбце `year_of_release` заменили модой по платформам, срок жизни которых совпадал с искомым периодом 2000-2013гг. Да, год релиза не будет соответствующим действительности, но зато игры, которые точно выходили в тот период, не будут утеряны. Остальные 158 пропусков было решено удалить.
- Два пропуска в `name` и `genre` решено игнорировать.
- После приведения значений в столбцах `name` и `genre` к одинаковому стилю написания и изменения значения "K-A" на "E" в столбце `rating_esrb` в датасете нашлось 237 строк-дубликатов, которые нужно было удалить.

---

## Фильтрация данных

Коллеги хотят изучить историю продаж игр в начале XXI века, и их интересует период с 2000 по 2013 год включительно. Отберите данные по этому показателю. Сохраните новый срез данных в отдельном датафрейме, например `df_actual`.

- Создаем срез с помощью метода .query:

In [52]:
df_actual = df.query('2000 <= year_of_release <= 2013')
df_actual['year_of_release'].unique()

<IntegerArray>
[2006, 2008, 2009, 2005, 2007, 2010, 2013, 2004, 2002, 2001, 2011, 2012, 2003,
 2000]
Length: 14, dtype: Int16

In [53]:
df_actual.info()

<class 'pandas.core.frame.DataFrame'>
Index: 12896 entries, 0 to 16954
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             12896 non-null  object 
 1   platform         12896 non-null  object 
 2   year_of_release  12896 non-null  Int16  
 3   genre            12896 non-null  object 
 4   na_sales         12896 non-null  float32
 5   eu_sales         12896 non-null  float32
 6   jp_sales         12896 non-null  float32
 7   other_sales      12896 non-null  float32
 8   critic_score     12896 non-null  Int8   
 9   user_score       12896 non-null  float32
 10  rating_esrb      12896 non-null  object 
dtypes: Int16(1), Int8(1), float32(5), object(4)
memory usage: 818.6+ KB


---

## Категоризация данных
    
Проведите категоризацию данных:
- Разделите все игры по оценкам пользователей и выделите такие категории: высокая оценка (от 8 до 10 включительно), средняя оценка (от 3 до 8, не включая правую границу интервала) и низкая оценка (от 0 до 3, не включая правую границу интервала).

Создаем категориальный столбец `user_score_category` с помощью метода .cut:

In [54]:
df_actual.loc[:,'user_score_category'] = pd.cut(df_actual.loc[:,'user_score'], 
                                                bins=[-1,0,3,8,10], 
                                                labels=['без оценки','низкая оценка','средняя оценка','высокая оценка'], 
                                                right=False)

C:\Users\kropi\AppData\Local\Temp\ipykernel_18412\3513226765.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_actual.loc[:,'user_score_category'] = pd.cut(df_actual.loc[:,'user_score'],


- Разделите все игры по оценкам критиков и выделите такие категории: высокая оценка (от 80 до 100 включительно), средняя оценка (от 30 до 80, не включая правую границу интервала) и низкая оценка (от 0 до 30, не включая правую границу интервала).

Создаем категориальный столбец `critic_score_category` с помощью метода .cut:

In [55]:
df_actual.loc[:,'critic_score_category'] = pd.cut(df_actual.loc[:,'critic_score'], 
                                                  bins=[-1,0,30,80,100], 
                                                  labels=['без оценки','низкая оценка','средняя оценка','высокая оценка'], 
                                                  right=False)

C:\Users\kropi\AppData\Local\Temp\ipykernel_18412\3169119872.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_actual.loc[:,'critic_score_category'] = pd.cut(df_actual.loc[:,'critic_score'],


In [56]:
df_actual.info()

<class 'pandas.core.frame.DataFrame'>
Index: 12896 entries, 0 to 16954
Data columns (total 13 columns):
 #   Column                 Non-Null Count  Dtype   
---  ------                 --------------  -----   
 0   name                   12896 non-null  object  
 1   platform               12896 non-null  object  
 2   year_of_release        12896 non-null  Int16   
 3   genre                  12896 non-null  object  
 4   na_sales               12896 non-null  float32 
 5   eu_sales               12896 non-null  float32 
 6   jp_sales               12896 non-null  float32 
 7   other_sales            12896 non-null  float32 
 8   critic_score           12896 non-null  Int8    
 9   user_score             12896 non-null  float32 
 10  rating_esrb            12896 non-null  object  
 11  user_score_category    12896 non-null  category
 12  critic_score_category  12896 non-null  category
dtypes: Int16(1), Int8(1), category(2), float32(5), object(4)
memory usage: 844.2+ KB


- После категоризации данных проверьте результат: сгруппируйте данные по выделенным категориям и посчитайте количество игр в каждой категории.

In [57]:
df_actual.groupby('user_score_category', observed=False)['name'].count()

user_score_category
без оценки        6361
низкая оценка      119
средняя оценка    4107
высокая оценка    2309
Name: name, dtype: int64

In [58]:
df_actual.groupby('critic_score_category', observed=False)['name'].count()

critic_score_category
без оценки        5659
низкая оценка       58
средняя оценка    5468
высокая оценка    1711
Name: name, dtype: int64

 У игроков больше и высоких, и низких оценок. Критики же более умеренные. Скорее всего это происходит из-за того, что в среднем игроки ставят оценки примерно так: понравилось - ставят 10, чтобы среднюю оценку поднять; не понравилось - ставят 1, чтобы опустить среднее арифметическое. А критики скорее именно оценивают игру по множеству показателей (хочется в это верить).

- Выделите топ-7 платформ по количеству игр, выпущенных за весь актуальный период.

In [59]:
df_actual.groupby('platform')['name'].count().sort_values(ascending=False).head(7)

platform
PS2     2161
DS      2150
Wii     1275
PSP     1196
X360    1121
PS3     1087
XB       824
Name: name, dtype: int64

PS2 и Nintendo DS лидируют с огромным отрывом.

---

## Итоговый вывод

В конце напишите основной вывод и отразите, какую работу проделали. Не забудьте указать описание среза данных и новых полей, которые добавили в исходный датасет.

 Работа проводилась над датасетом `new_games.csv`, который содержит 16956 строк и 11 столбцов. В датасете была размещена информация об играх. Знакомство с данными показало следующее:
- Два пропуска в столбцах `name` и `genre`, 275 пропусков в столбце `year_of_release` и от 40% до 51% пропусков в столбцах `critic_score`, `user_score` и `rating_esrb`.
- Так же для удобства были изменены названия столбцов и типы данных:
    - Названия приведены к стилю snake case, а так же изменено название столбца `Rating` на `rating_esrb` для лучшей читаемости.
    - `eu_sales`, `jp_ sales` и `user_score` приведены к типу `float32` с заменой неподходящих значений на пропуски.
    - `critic_score` приведен к типу `Int8`, `year_of_release` к типу `Int16`.
- Для выполнения задачи клиента было сделао следующее:
    - Создан срез изначального датафрейма по столбцу `year_of_release` включающий значения от 2000 до 2013 включительно и сохранен в новый датафрейм `df_actual`.
    - В новом датафрейме созданы новые столбцы `user_score_category` и `critic_score_category`. Это категориальные столбцы со значениями, основанными на оценке игроков и критиков соответственно.
- Далее по заданию были сделаны группировки по значениям столбцов `user_score_category` и `critic_score_category` с агрегацией по количеству игр в каждой категории.
- А так же выведен топ-7 платформ по количеству игр.